atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/backup/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,222 @@
+        // resource, subsequent #runWithResource calls would now block on acquiring the lock. This would be doubly
+        // damaging if the resource lock was fair, as now #runWithResource calls are blocked until all calls prior to
+        // closing the resource have finished.
+
+        // To avoid this, we must not block an update with close. This can either be done with another thread, or
+        // simply using a subscriber on the updated resource.
+        refreshableSubscriptionDisposable = refreshableResourceWithLock.subscribe(resourceWithLock -> {
+            Optional<Pair<ReadWriteLock, T>> maybeResourceWithLockToClose =
+                    currentResourceWithLock.getAndSet(Optional.of(resourceWithLock));
+            maybeResourceWithLockToClose.ifPresent(this::closeResourceAndLock);
+        });
+    }
+
+    private void closeResourceAndLock(Pair<ReadWriteLock, T> resourceWithLockToClose) {
+        Lock writeLock = resourceWithLockToClose.getLhSide().writeLock();
+        runWithLock(writeLock, () -> {
+            try {
+                T resourceToClose = resourceWithLockToClose.getRhSide();
+                resourceToClose.close();
+            } catch (IOException e) {
+                log.warn("Failed to close resource. This may result in a resource leak", e);
+            }
+        });
+    }
+
+    /***
+     * Closes the container and the latest created resource.
+     * After this method has returned, no further resources will be created in this container, and all subsequent
+     * {@link #runWithResource} invocations will fail with a {@link SafeIllegalStateException}.
+     *
+     * Note: This method does <b>block</b> on any inflight {@link #runWithResource} invocations.
+     *
+     * Any {@link IOException} from closing the underlying resource will not be propagated.
+     *
+     * See {@link Closeable#close()}.
+     */
+    @Override
+    public void close() {
+        Lock isClosedWriteLock = isClosedLock.writeLock();
+        runWithLock(isClosedWriteLock, () -> {
+            isClosed = true;
+            currentResourceWithLock.get().ifPresent(this::closeResourceAndLock);
+            refreshableSubscriptionDisposable.dispose();
+        });
+    }
+
+    /***
+     * Returns the result of `task` applied with the current resource.
+     * This resource is guaranteed not to be closed within scope of this function <i>by this container</i>.
+     * There are no guarantees on the resource being closed externally e.g. if there is a shared reference, nor if the
+     * resource is closed within scope by the task itself.
+     *
+     * If an update to the refreshable occurs during an invocation, any <i>inflight</i> invocations will <i>not</i> use
+     * the new resource.
+     *
+     * <b>Do not</b> block on close or update the refreshable within the scope of runWith. Doing so will result in a
+     * deadlock.
```

Comment:
Pointing some attention to this - getting into this circumstance would be a rather weird use of this API, but it's still possible.